docs/source/type_narrowing.rst

```diff
@@ -83,6 +84,38 @@ We can also use ``assert`` to narrow types in the same context:
      reveal_type(x)  # Revealed type is "builtins.int"
      print(x + '!')  # Typechecks with `mypy`, but fails in runtime.
 
+
+Limitations of narrowing
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Mypy's analysis is limited to individual symbols and it will not track
+relationships between symbols. For example, in the following code
+it's easy to deduce that if :code:`a` is None then :code:`b` must not be,
+therefore :code:`a or b` will always be a string, but Mypy will not be able to tell that:
+
+.. code-block:: python
+
+    def f(a: str | None, b: str | None) -> str:
+        if a is not None or b is not None:
+            return a or b  # Incompatible return value type (got "str | None", expected "str")
+        return 'spam'
+
+Tracking these sort of cross-variable conditions in a type checker would add significant complexity
+and performance overhead and would be computationally infeasible in all but the most basic cases.
+
+You may override the type checker with a :ref:`cast <casts>`, or rewrite the function to be
```

Comment:
```suggestion
You may override the type checker with an ``assert`` or :ref:`cast <casts>`, or rewrite the function to be
```
When I run into this, assert is usually more ergonomic. It's also safer than cast.

Reply:
Probably would phrase as "help the type checker with an ``assert``", but I'm not sure it'll be more ergonomic in this case. Maybe I'm tired and missing something obvious?

cast:
```py
def f(a: str | None, b: str | None) -> str:
    if a is not None or b is not None:
        return a or cast(str, b)
    return 'spam'
```

assert:
```py
def f(a: str | None, b: str | None) -> str:
    if a is not None or b is not None:
        if a:
		   return a
		assert b
		return b
    return 'spam'
```

third option:
```py
def f(a: str | None, b: str | None) -> str:
    if a is not None or b is not None:
        return a or b or 'impossible'
    return 'spam'
```

Reply:
Yeah, not more ergonomic in this case, but better in the cases where I usually run into this. But maybe I dislike cast more than most people :-)

Reply:
`assert a or b` doesn't help you either because we don't narrow conditionally. (Also, in this case you'd get more than you'd bargain for, as empty `b` would assert.)

Reply:
Oh oops, I guess you'd have to do `ret = a or b; assert ret is not None; return ret`